### PR TITLE
checker(dm): port free space lightning precheck

### PR DESCRIPTION
**Pull request by lance6716**
Signed-off-by: lance6716 <lance6716@gmail.com>

### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tiflow/issues/3510

### What is changed and how it works?

get source table size by `SELECT data_length + index_length FROM information_schema.TABLES WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ?` and use it in lightning prechecks

#### Tests 

 - Integration test

### Release note 

```release-note
`None`.
```

### CODE REVIEW

File: dm/checker/checker.go [hunk 1]
```diff
@@ -154,55 +167,98 @@ func (c *Checker) Init(ctx context.Context) (err error) {
...
 	if egErr := eg.Wait(); egErr != nil {
-		return egErr
+		return nil, egErr
+	}
+
+	info.totalDataSize = atomic.NewInt64(0)
+	if _, ok := c.checkingItems[config.LightningFreeSpaceChecking]; ok &&
+		c.stCfgs[0].LoaderConfig.ImportMode == config.LoadModePhysical &&
+		c.stCfgs[0].Mode != config.ModeIncrement {
+		// TODO: concurrently read it intra-source later
```
Comment (lance6716) [inline dm/checker/checker.go:177]: this PR is a bit large, I prefer fix it in next PR

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: dm/checker/checker.go [hunk 2]
```diff
@@ -115,23 +119,32 @@ func NewChecker(cfgs []*config.SubTaskConfig, checkingItems map[string]string, e
...
-	rollbackHolder.Add(fr.FuncRollback{Name: "close-DBs", Fn: c.closeDBs})
-
-	c.tctx = tcontext.NewContext(ctx, log.With(zap.String("unit", "task check")))
-
-	// 1. get allow-list of tables and routed table name from upstream and downstream
+// tablePairInfo records information about a upstream-downstream table pair including
+// members may have repeated meanings but have different data structure to satisfy different usages.
+type tablePairInfo struct {
+	// downstream table -> sourceID -> upstream tables
+	tablesPerDownstreamTable map[filter.Table]map[string][]filter.Table
```
[the remark this answers is not in the document]
Comment (lance6716) [reply dm/checker/checker.go:126]: there's two name type: source/target and upstream/downstream. Do you prefer we always use source/target?

File: dm/config/checking_item_test.go [hunk 3]
```diff
@@ -16,28 +16,30 @@ package config
...
+		require.NoError(t, ValidateCheckingItem(item))
+		if slices.Contains(LightningPrechecks, item) {
+			lightningCheck++
+		} else {
+			normalCheck++
+		}
 	}
-	c.Assert(ValidateCheckingItem("xxx"), NotNil)
+	// remember to update the number when add new checking items.
+	require.Equal(t, 5, lightningCheck)
```
Comment (D3Hunter) [inline dm/config/checking_item_test.go:34]: this doesn't check `LightningPrechecks` ⊆ `AllCheckingItems`

File: dm/checker/checker.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (D3Hunter) [reply dm/checker/checker.go:126]: prefer source/target, they have less chars 😅, upstream/downstream are ok too

**Review by D3Hunter — APPROVED** [review]

File: dm/config/task.go [hunk 4]
```diff
@@ -326,6 +329,10 @@ func (m *LoaderConfig) adjust() error {
 		m.PoolSize = defaultPoolSize
 	}
 
+	if m.Dir != "" && m.SortingDirPhysical == "" {
+		m.SortingDirPhysical = m.Dir
```
Comment (buchuitoudegou) [inline dm/config/task.go:333]: Why put the `sorting-dir` with `source-dir`? Lightning will prefer they are in different disk. Maybe we can set another default `sorting-dir`?

File: dm/pkg/checker/lightning.go [hunk 5]
```diff
@@ -135,3 +138,107 @@ func (c *LightningClusterVersionChecker) Check(ctx context.Context) *Result {
...
+	}
+	clusterAvail := uint64(0)
+	for _, store := range storeInfo.Stores {
+		clusterAvail += uint64(store.Status.Available)
+	}
+	if clusterAvail < uint64(c.sourceDataSize) {
+		result.State = StateFailure
+		result.Errors = append(result.Errors, &Error{
+			Severity: StateFailure,
+			ShortErr: fmt.Sprintf("Cluster doesn't have enough space, available is %s, but we need %s",
```
Comment (buchuitoudegou) [inline dm/pkg/checker/lightning.go:181]:
```suggestion
			ShortErr: fmt.Sprintf("Downstream doesn't have enough space, available is %s, but we need %s",
```

**Review by buchuitoudegou — COMMENTED** [review]
rest lgtm

File: dm/config/task.go (hunk 4 above)
Comment (lance6716) [reply dm/config/task.go:333]: I don't know how to do it in code 😂 and user may be unexpected to see another disk is occupied

Comment (buchuitoudegou) [reply dm/config/task.go:333]: 👌 Maybe just leave some comments to demonstrate the risk here?

**Review by buchuitoudegou — APPROVED** [review]

**Comment by lance6716** [comment]
/hold

I prefer we merge it after v6.5 branch is created

**Pull request merged**